text-emphasis marks should not be rendered if there is no emphasized character

vitorroriz opened the pull request:
#### 3b04a522163d5f16917facb89c73d5686d3de884
<pre>
text-emphasis marks should not be rendered if there is no emphasized character
<a href="https://bugs.webkit.org/show_bug.cgi?id=251201">https://bugs.webkit.org/show_bug.cgi?id=251201</a>
rdar://problem/104688963

Reviewed by Myles C. Maxfield.

There were 2 different reasons for failing these tests:
1. When drawing emphasis mark (drawEmphasisMarks())
we weren&apos;t skipping the glyphs that were replaced
by the deleted glyph (0xFFFF).

2. When iterating through the characters to be drawn with
the text iterators, we weren&apos;t converting characters encoded
as surrogated pairs (UTF16). When one of these characters
would show up, we would pass them to canReceiveTextEmphasis(),
implicitly converting the UChar word to a UChar32 word,
without properly converting the surrogated pairs.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::drawEmphasisMarks const):
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::makeGlyphInvisible):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
(WebCore::WidthIterator::applyCSSVisibilityRules):
* Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:

Canonical link: <a href="https://commits.webkit.org/262997@main">https://commits.webkit.org/262997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17d05d4f1c482d073a71f8b2f3a8159b25e38c72

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3210 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3275 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3386 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4627 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3561 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3181 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3303 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2794 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3245 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3556 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2893 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4439 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1057 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2864 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2773 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2837 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2916 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4189 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3290 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2625 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2865 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2866 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/796 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2862 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3128 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -299,8 +300,13 @@ inline void WidthIterator::advanceInternal(TextIterator& textIterator, GlyphBuff
+        if (m_forTextEmphasis) {
+            UChar32 ch32 = character;
+            if (U16_IS_SURROGATE(character))
+                U16_GET(m_run.characters16(), 0, currentCharacterIndex, m_run.length(), ch32);
```
U16_GET should already handle both surrogates and non-surrogates. So I think, instead of conditionally calling it here, we should hoist it up and unconditionally call it to produce the `character` variable.

vitorroriz replied:
I've just realized that `WidthIterator::advance()` will use as iterator `ComposedCharacterClusterTextIterator` or `SurrogatePairAwareTextIterator` for the non-8-bit case. Both already use `U16_NEXT()` when producing character (::consume()), so I think we don't have to call U16_GET at all for WidthIterator.

litherum commented on Source/WebCore/platform/graphics/WidthIterator.cpp:
```diff
@@ -299,8 +300,13 @@ inline void WidthIterator::advanceInternal(TextIterator& textIterator, GlyphBuff
+            if (U16_IS_SURROGATE(character))
+                U16_GET(m_run.characters16(), 0, currentCharacterIndex, m_run.length(), ch32);
+            if (!FontCascade::canReceiveTextEmphasis(character))
+                glyph = 0;
```
Are you sure you don't want this to be 0xFFFF? 0 seems wrong.

vitorroriz replied:
In ComplexIterator we were setting it to 0 before this PR. IIRC it is later replaced by 0xFFFF, so I can try to set it directly to 0xFFFF.

vitorroriz committed:
text-emphasis marks should not be rendered if there is no emphasized character
https://bugs.webkit.org/show_bug.cgi?id=251201
rdar://problem/104688963

Reviewed by Myles C. Maxfield.

There were 2 different reasons for failing these tests:
1. When drawing emphasis mark (drawEmphasisMarks())
we weren't skipping the glyphs that were replaced
by the deleted glyph (0xFFFF).

2. When iterating through the characters to be drawn with
the text iterators, we weren't converting characters encoded
as surrogated pairs (UTF16). When one of these characters
would show up, we would pass them to canReceiveTextEmphasis(),
implicitly converting the UChar word to a UChar32 word,
without properly converting the surrogated pairs.

* LayoutTests/TestExpectations:
* Source/WebCore/platform/graphics/ComplexTextController.cpp:
(WebCore::ComplexTextController::adjustGlyphsAndAdvances):
* Source/WebCore/platform/graphics/FontCascade.cpp:
(WebCore::FontCascade::drawEmphasisMarks const):
* Source/WebCore/platform/graphics/GlyphBuffer.h:
(WebCore::GlyphBuffer::makeGlyphInvisible):
* Source/WebCore/platform/graphics/WidthIterator.cpp:
(WebCore::WidthIterator::advanceInternal):
(WebCore::WidthIterator::applyCSSVisibilityRules):
* Source/WebCore/platform/graphics/coretext/GlyphPageCoreText.cpp:

Canonical link: https://commits.webkit.org/262997@main
